# Introduction

In this notebook, we will do a comprehensive analysis of the Android app market by comparing thousands of apps in the Google Play store.

# About the Dataset of Google Play Store Apps & Reviews

**Data Source:** <br>
App and review data was scraped from the Google Play Store by Lavanya Gupta in 2018. Original files listed [here](
https://www.kaggle.com/lava18/google-play-store-apps).

# Import Statements

In [1]:
import pandas as pd
import plotly.express as px

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# Notebook Presentation

In [3]:
# Show numeric output in decimal format e.g., 2.15
pd.options.display.float_format = '{:,.2f}'.format

# Read the Dataset

In [4]:
df_apps = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Google+Play+Store+Project+(Start) (1)/apps.csv')

# Data Cleaning

We will answer these questions: 
- How many rows and columns does `df_apps` have? 
- What are the column names? 
- Look at a random sample of 5 different rows with [.sample()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html).

In [5]:
print(df_apps.shape)
df_apps.sample(5)

(10841, 12)


App             Category  \
2917                          Learn CT Scan Of Head  BOOKS_AND_REFERENCE   
2111                                FB Photographie          PHOTOGRAPHY   
3675  BBW Dating & Curvy Singles Chat- LargeFriends               DATING   
4752                            WSVN - 7 News Miami   NEWS_AND_MAGAZINES   
7470                                        Ao Oni2                 GAME   

      Rating  Reviews  Size_MBs   Installs  Type Price Content_Rating  \
2917    4.10       10      5.70      5,000  Free     0       Everyone   
2111    4.70       37     10.00      1,000  Free     0       Everyone   
3675    4.40      218     27.00     10,000  Free     0     Mature 17+   
4752    4.50     3320     14.00    100,000  Free     0       Everyone   
7470    4.10    21223     48.00  1,000,000  Free     0           Teen   

                 Genres    Last_Updated Android_Ver  
2917  Books & Reference   July 22, 2018  4.1 and up  
2111        Photography    May 22, 2016  4.0 and up  
3675             Dating   July 23, 2018  4.1 and up  
4752   News & Magazines  August 3, 2018  5.0 and up  
7470          Adventure  April 20, 2018  4.1 and up

### Drop Unused Columns

We remove the columns called `Last_Updated` and `Android_Version` from the DataFrame. We will not use these columns. 

In [6]:
df_apps.drop(axis =1, columns = ["Last_Updated", "Android_Ver"], inplace = True)
print(df_apps.shape)
df_apps.head(5)

(10841, 10)


App         Category  Rating  \
0                       Ak Parti Yardım Toplama           SOCIAL     NaN   
1                    Ain Arabic Kids Alif Ba ta           FAMILY     NaN   
2  Popsicle Launcher for Android P 9.0 launcher  PERSONALIZATION     NaN   
3                     Command & Conquer: Rivals           FAMILY     NaN   
4                                    CX Network         BUSINESS     NaN   

   Reviews  Size_MBs Installs  Type   Price Content_Rating           Genres  
0        0      8.70        0  Paid  $13.99           Teen           Social  
1        0     33.00        0  Paid   $2.99       Everyone        Education  
2        0      5.50        0  Paid   $1.49       Everyone  Personalization  
3        0     19.00        0   NaN       0   Everyone 10+         Strategy  
4        0     10.00        0  Free       0       Everyone         Business

### Find and Remove NaN values in Ratings

**Challenge**: How may rows have a NaN value (not-a-number) in the Ratings column? Create DataFrame called `df_apps_clean` that does not include these rows. 

In [7]:
df_apps['Rating'].isna().sum()

1474

There are 1474 NaNs in the Ratign Columns

In [8]:
df_apps_clean = df_apps.dropna()
df_apps_clean.shape

(9367, 10)

### Find and Remove Duplicates

We will answer these questions: 

- Are there any duplicates in data? Check for duplicates using the [.duplicated()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html) function.
- How many entries can you find for the "Instagram" app? Use [.drop_duplicates()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html) to remove any duplicates from `df_apps_clean`. 


In [9]:
duplicated_rows = df_apps_clean[df_apps_clean.duplicated()]
print(duplicated_rows.shape)
duplicated_rows.head()

(476, 10)


App Category  Rating  Reviews  \
946                       420 BZ Budeze Delivery  MEDICAL    5.00        2   
1133                                 MouseMingle   DATING    2.70        3   
1196  Cardiac diagnosis (heart rate, arrhythmia)  MEDICAL    4.40        8   
1231                                Sway Medical  MEDICAL    5.00        3   
1247              Chat Kids - Chat Room For Kids   DATING    4.70        6   

      Size_MBs Installs  Type   Price Content_Rating   Genres  
946      11.00      100  Free       0     Mature 17+  Medical  
1133      3.90      100  Free       0     Mature 17+   Dating  
1196      6.50      100  Paid  $12.99       Everyone  Medical  
1231     22.00      100  Free       0       Everyone  Medical  
1247      4.90      100  Free       0     Mature 17+   Dating

In [10]:
df_apps_clean[df_apps_clean.App == "Instagram"]

App Category  Rating   Reviews  Size_MBs       Installs  Type  \
10806  Instagram   SOCIAL    4.50  66577313      5.30  1,000,000,000  Free   
10808  Instagram   SOCIAL    4.50  66577446      5.30  1,000,000,000  Free   
10809  Instagram   SOCIAL    4.50  66577313      5.30  1,000,000,000  Free   
10810  Instagram   SOCIAL    4.50  66509917      5.30  1,000,000,000  Free   

      Price Content_Rating  Genres  
10806     0           Teen  Social  
10808     0           Teen  Social  
10809     0           Teen  Social  
10810     0           Teen  Social

WOW, there are a lots of apps have the name "Instagram". Let's drop those app have the same name.

In [11]:
df_apps_clean = df_apps_clean.drop_duplicates()
df_apps_clean[df_apps_clean.App == "Instagram"]

App Category  Rating   Reviews  Size_MBs       Installs  Type  \
10806  Instagram   SOCIAL    4.50  66577313      5.30  1,000,000,000  Free   
10808  Instagram   SOCIAL    4.50  66577446      5.30  1,000,000,000  Free   
10810  Instagram   SOCIAL    4.50  66509917      5.30  1,000,000,000  Free   

      Price Content_Rating  Genres  
10806     0           Teen  Social  
10808     0           Teen  Social  
10810     0           Teen  Social

Hmm, Maybe the default is too strict. because there are entries of the same app with different number of reviews are still being kept. Let's not judge duplicates base on all columns but only `App`, `Type`, and `Price`. 

In [12]:
df_apps_clean = df_apps_clean.drop_duplicates(subset = ["App", "Type", "Price"])
df_apps_clean[df_apps_clean.App == "Instagram"]

App Category  Rating   Reviews  Size_MBs       Installs  Type  \
10806  Instagram   SOCIAL    4.50  66577313      5.30  1,000,000,000  Free   

      Price Content_Rating  Genres  
10806     0           Teen  Social

Yea, it seems more right here.

In [13]:
df_apps_clean.shape

(8199, 10)

Ok, so that leave us only 8199 entries and 10 columns after the cleaning.

# Find Highest Rated Apps

We will identify which apps are the highest rated. What problem might you encounter if you rely exclusively on ratings alone to determine the quality of an app?

In [14]:
highest_rated_apps = df_apps_clean[df_apps_clean.Rating == df_apps_clean.Rating.max()]
highest_rated_apps.head()

App Category  Rating  Reviews  \
21                              KBA-EZ Health Guide  MEDICAL    5.00        4   
28                                         Ra Ga Ba     GAME    5.00        2   
47                                          Mu.F.O.     GAME    5.00        2   
82                                 Brick Breaker BR     GAME    5.00        7   
99  Anatomy & Physiology Vocabulary Exam Review App  MEDICAL    5.00        1   

    Size_MBs Installs  Type  Price Content_Rating   Genres  
21     25.00        1  Free      0       Everyone  Medical  
28     20.00        1  Paid  $1.49       Everyone   Arcade  
47     16.00        1  Paid  $0.99       Everyone   Arcade  
82     19.00        5  Free      0       Everyone   Arcade  
99      4.60        5  Free      0       Everyone  Medical

These apps have a really high ratings, but if we look at the `Reviews` columns, they only have a few reviews (fewer than 10). Relying exclusively on ratings alone to determine the quality of an app.

# Find 5 Largest Apps in terms of Size (MBs)

We'll answer the following question:
- What's the size in megabytes (MB) of the largest Android apps in the Google Play Store. 
-Based on the data, do you think there could be limit in place or can developers make apps as large as they please? 

In [15]:
highest_size_apps = df_apps_clean[df_apps_clean.Size_MBs == df_apps_clean.Size_MBs.max()]
highest_size_apps.head()

App            Category  \
1795                              Navi Radiography Pro             MEDICAL   
3144                                        Vi Trainer  HEALTH_AND_FITNESS   
4176  Car Crash III Beam DH Real Damage Simulator 2018                GAME   
7926                                         Post Bank             FINANCE   
7927                       The Walking Dead: Our World                GAME   

      Rating  Reviews  Size_MBs   Installs  Type   Price Content_Rating  \
1795    4.70       11    100.00        500  Paid  $15.99       Everyone   
3144    3.60      124    100.00      5,000  Free       0       Everyone   
4176    3.60      151    100.00     10,000  Free       0       Everyone   
7926    4.50    60449    100.00  1,000,000  Free       0       Everyone   
7927    4.00    22435    100.00  1,000,000  Free       0           Teen   

                Genres  
1795           Medical  
3144  Health & Fitness  
4176            Racing  
7926           Finance  
7927            Action

Here we can clearly see that there seems to be an upper bound of 100 MB for the size of an app. A quick google search would also have revealed that this limit is imposed by the Google Play Store itself. It’s interesting to see that a number of apps actually hit that limit exactly.

# Find the 5 App with Most Reviews

We'll answer the following questions: 
- Which apps have the highest number of reviews? 
- Are there any paid apps among the top 50?

In [16]:
highest_review_apps = df_apps_clean.sort_values(by = ["Reviews"], ascending = False)
highest_review_apps.head(50)

App            Category  \
10805                                           Facebook              SOCIAL   
10785                                 WhatsApp Messenger       COMMUNICATION   
10806                                          Instagram              SOCIAL   
10784           Messenger – Text and Video Chat for Free       COMMUNICATION   
10650                                     Clash of Clans                GAME   
10744            Clean Master- Space Cleaner & Antivirus               TOOLS   
10835                                     Subway Surfers                GAME   
10828                                            YouTube       VIDEO_PLAYERS   
10746  Security Master - Antivirus, VPN, AppLock, Boo...               TOOLS   
10584                                       Clash Royale                GAME   
10763                                   Candy Crush Saga                GAME   
10770        UC Browser - Fast Download Private & Secure       COMMUNICATION   
10735                                           Snapchat              SOCIAL   
10489    360 Security - Free Antivirus, Booster, Cleaner               TOOLS   
10731                                     My Talking Tom                GAME   
10594                                        8 Ball Pool                GAME   
10302  DU Battery Saver - Battery Charger & Battery Life               TOOLS   
10354                        BBM - Free Calls & Messages       COMMUNICATION   
10549  Cache Cleaner-DU Speed Booster (booster & clea...               TOOLS   
10757                                            Twitter  NEWS_AND_MAGAZINES   
10721                                    Viber Messenger       COMMUNICATION   
10578                                     Shadow Fight 2                GAME   
10813                                      Google Photos         PHOTOGRAPHY   
10724                        LINE: Free Calls & Messages       COMMUNICATION   
10717                                                Pou                GAME   
10792                      Skype - free IM & video calls       COMMUNICATION   
10628                                         Pokémon GO                GAME   
10388           Minion Rush: Despicable Me Official Game                GAME   
10694                                            Yes day                GAME   
10695                                            Hay Day              FAMILY   
10644                           Dream League Soccer 2018                GAME   
10696                                  My Talking Angela                GAME   
10660             VivaVideo - Video Editor & Photo Movie       VIDEO_PLAYERS   
10786                       Google Chrome: Fast & Secure       COMMUNICATION   
10817                          Maps - Navigate & Explore    TRAVEL_AND_LOCAL   
10672                                  Hill Climb Racing                GAME   
10734                                      Facebook Lite              SOCIAL   
10649                                Asphalt 8: Airborne                GAME   
10699                          Mobile Legends: Bang Bang                GAME   
10322  Battery Doctor-Battery Life Saver & Battery Co...               TOOLS   
10396                                     Piano Tiles 2™                GAME   
10777                                       Temple Run 2                GAME   
10822                                             Google               TOOLS   
10359  Truecaller: Caller ID, SMS spam blocking & Dialer       COMMUNICATION   
10711                         SHAREit - Transfer & Share               TOOLS   
10389   Sniper 3D Gun Shooter: Free Shooting Games - FPS                GAME   
10676                                   Farm Heroes Saga                GAME   
10576   PicsArt Photo Studio: Collage Maker & Pic Editor         PHOTOGRAPHY   
10461  PhotoGrid: Video & Pic Collage Maker, Photo Ed...         PHOTOGRAPHY   
10502   GO Launcher - 3D parallax Themes & HD Wallpap

If you look at the number of reviews, you can find the most popular apps on the Android App Store. These include the usual suspects: Facebook, WhatsApp, Instagram etc. What’s also notable is that the list of the top 50 most reviewed apps does not include a single paid app!

# Plotly Pie and Donut Charts - Visualise Categorical Data: Content Ratings

All Android apps have a content rating like “Everyone” or “Teen” or “Mature 17+”. Let’s take a look at the distribution of the content ratings in our dataset and see how to visualise it with [plotly](https://plotly.com/python/) - a popular data visualisation library that you can use alongside or instead of Matplotlib.

## Count the numbers of Ratings

In [17]:
ratings = df_apps_clean.Content_Rating.value_counts()
ratings

Everyone           6621
Teen                912
Mature 17+          357
Everyone 10+        305
Adults only 18+       3
Unrated               1
Name: Content_Rating, dtype: int64

In [18]:
fig = px.pie(labels = ratings.index, 
             values = ratings.values,
             )
fig.show()

Let's make it more beautiful

In [19]:
fig = px.pie(labels = ratings.index, 
             values = ratings.values,
             title = "Content Ratings",
             names = ratings.index)

fig.update_traces(textposition='outside', textinfo='percent+label')

fig.show()

How about creating a donut chart? 

In [20]:
fig = px.pie(labels=ratings.index,
values=ratings.values,
title="Content Rating",
names=ratings.index,
hole=0.6,
)
fig.update_traces(textposition='inside', textfont_size=15, textinfo='percent')
 
fig.show()


# Numeric Type Conversion: Examine the Number of Installs
We will answer the folloing questions: 
- How many apps had over 1 billion (that's right - BILLION) installations? 
- How many apps just had a single install? 

- Check the datatype of the Installs column.

- Count the number of apps at each level of installations. 

- Convert the number of installations (the Installs column) to a numeric data type. This is a 2-step process. make sure you remove non-numeric characters first. 

In [21]:
df_apps_clean.Installs.describe()

count          8199
unique           19
top       1,000,000
freq           1417
Name: Installs, dtype: object

In [22]:
df_apps_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8199 entries, 21 to 10835
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             8199 non-null   object 
 1   Category        8199 non-null   object 
 2   Rating          8199 non-null   float64
 3   Reviews         8199 non-null   int64  
 4   Size_MBs        8199 non-null   float64
 5   Installs        8199 non-null   object 
 6   Type            8199 non-null   object 
 7   Price           8199 non-null   object 
 8   Content_Rating  8199 non-null   object 
 9   Genres          8199 non-null   object 
dtypes: float64(2), int64(1), object(7)
memory usage: 704.6+ KB


Both of these show that we are dealing with a non-numeric data type. In this case, the type is "object".

Let's explore it further!

In [23]:
df_apps_clean[["App", "Installs"]].groupby("Installs").count()

App
Installs           
1                 3
1,000           698
1,000,000      1417
1,000,000,000    20
10               69
10,000          988
10,000,000      933
100             303
100,000        1096
100,000,000     189
5                 9
5,000           425
5,000,000       607
50               56
50,000          457
50,000,000      202
500             199
500,000         504
500,000,000      24

First, let's remove the comma ",".

In [24]:
df_apps_clean.Installs = df_apps_clean.Installs.astype(str).str.replace(',', "")

Then convert them to numberic type. 

In [25]:
df_apps_clean.Installs = pd.to_numeric(df_apps_clean.Installs)
df_apps_clean[['App', 'Installs']].groupby('Installs').count()

App
Installs        
1              3
5              9
10            69
50            56
100          303
500          199
1000         698
5000         425
10000        988
50000        457
100000      1096
500000       504
1000000     1417
5000000      607
10000000     933
50000000     202
100000000    189
500000000     24
1000000000    20

# Find the Most Expensive Apps, Filter out the Junk, and Calculate a (ballpark) Sales Revenue Estimate

Let's examine the Price column more closely.

We will do the following: 

- Convert the price column to numeric data. Then investigate the top 20 most expensive apps in the dataset.

- Remove all apps that cost more than $250 from the `df_apps_clean` DataFrame.

- Add a column called 'Revenue_Estimate' to the DataFrame. This column should hold the price of the app times the number of installs. What are the top 10 highest grossing paid apps according to this estimate? Out of the top 10 highest grossing paid apps, how many are games?


In [26]:
df_apps_clean.Price = df_apps_clean.Price.astype(str).str.replace('$', "")
df_apps_clean.Price = pd.to_numeric(df_apps_clean.Price)
expensive_apps = df_apps_clean.sort_values(by = "Price", ascending = False)
expensive_apps.head(20)

<ipython-input-26-a663efbe5fc1>:1: FutureWarning:

The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.



App   Category  Rating  Reviews  Size_MBs  \
3946        I'm Rich - Trump Edition  LIFESTYLE    3.60      275      7.30   
2461              I AM RICH PRO PLUS    FINANCE    4.00       36     41.00   
4606               I Am Rich Premium    FINANCE    4.10     1867      4.70   
3145              I am rich(premium)    FINANCE    3.50      472      0.94   
3554                      💎 I'm rich  LIFESTYLE    3.80      718     26.00   
5765                       I am rich  LIFESTYLE    3.80     3547      1.80   
1946  I am rich (Most expensive app)    FINANCE    4.10      129      2.70   
2775                   I Am Rich Pro     FAMILY    4.40      201      2.70   
3221                  I am Rich Plus     FAMILY    4.00      856      8.70   
3114                       I am Rich    FINANCE    4.30      180      3.80   
1331          most expensive app (H)     FAMILY    4.30        6      1.50   
2394                      I am Rich!    FINANCE    3.80       93     22.00   
3897                       I Am Rich     FAMILY    3.60      217      4.90   
2193             I am extremely Rich  LIFESTYLE    2.90       41      2.90   
3856                   I am rich VIP  LIFESTYLE    3.80      411      2.60   
2281       Vargo Anesthesia Mega App    MEDICAL    4.60       92     32.00   
1407                    LTC AS Legal    MEDICAL    4.00        6      1.30   
2629                I am Rich Person  LIFESTYLE    4.20      134      1.80   
2481         A Manual of Acupuncture    MEDICAL    3.50      214     68.00   
4264         Golfshot Plus: Golf GPS     SPORTS    4.10     3387     25.00   

      Installs  Type  Price Content_Rating         Genres  
3946     10000  Paid 400.00       Everyone      Lifestyle  
2461      1000  Paid 399.99       Everyone        Finance  
4606     50000  Paid 399.99       Everyone        Finance  
3145      5000  Paid 399.99       Everyone        Finance  
3554     10000  Paid 399.99       Everyone      Lifestyle  
5765    100000  Paid 399.99       Everyone      Lifestyle  
1946      1000  Paid 399.99           Teen        Finance  
2775      5000  Paid 399.99       Everyone  Entertainment  
3221     10000  Paid 399.99       Everyone  Entertainment  
3114      5000  Paid 399.99       Everyone        Finance  
1331       100  Paid 399.99       Everyone  Entertainment  
2394      1000  Paid 399.99       Everyone        Finance  
3897     10000  Paid 389.99       Everyone  Entertainment  
2193      1000  Paid 379.99       Everyone      Lifestyle  
3856     10000  Paid 299.99       Everyone      Lifestyle  
2281      1000  Paid  79.99       Everyone        Medical  
1407       100  Paid  39.99       Everyone        Medical  
2629      1000  Paid  37.99       Everyone      Lifestyle  
2481      1000  Paid  33.99       Everyone        Medical  
4264     50000  Paid  29.99       Everyone         Sports

I thought I did something wrong...but it's actually true that those 20 most expensive apps are actually expensive. There are 15 I am Rich Apps in the Google Play Store apparently. They all cost $300 or more, which is the main point of the app. LOLLL

Leaving this bad data in our dataset will misrepresent our analysis of the most expensive 'real' apps. Let's remove apps that are more than $250

### The most expensive apps sub $250

In [27]:
df_apps_clean = df_apps_clean[df_apps_clean['Price'] < 250]
df_apps_clean.sort_values('Price', ascending=False).head(5)

App   Category  Rating  Reviews  Size_MBs  \
2281  Vargo Anesthesia Mega App    MEDICAL    4.60       92     32.00   
1407               LTC AS Legal    MEDICAL    4.00        6      1.30   
2629           I am Rich Person  LIFESTYLE    4.20      134      1.80   
2481    A Manual of Acupuncture    MEDICAL    3.50      214     68.00   
2463         PTA Content Master    MEDICAL    4.20       64     41.00   

      Installs  Type  Price Content_Rating     Genres  
2281      1000  Paid  79.99       Everyone    Medical  
1407       100  Paid  39.99       Everyone    Medical  
2629      1000  Paid  37.99       Everyone  Lifestyle  
2481      1000  Paid  33.99       Everyone    Medical  
2463      1000  Paid  29.99       Everyone    Medical

We look at the top 5 apps now, we see that 4 out of 5 are medical apps.

In [28]:
df_apps_clean["Revenue_Estimate"] = df_apps_clean["Price"]*df_apps_clean["Installs"]
df_apps_clean.sort_values('Revenue_Estimate', ascending=False)[:10]


<ipython-input-28-66d9c8e78ed6>:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



App     Category  Rating  Reviews  Size_MBs  \
9220                      Minecraft       FAMILY    4.50  2376564     19.00   
8825                  Hitman Sniper         GAME    4.60   408292     29.00   
7151  Grand Theft Auto: San Andreas         GAME    4.40   348962     26.00   
7477            Facetune - For Free  PHOTOGRAPHY    4.40    49553     48.00   
7977        Sleep as Android Unlock    LIFESTYLE    4.50    23966      0.85   
6594            DraStic DS Emulator         GAME    4.60    87766     12.00   
6082                   Weather Live      WEATHER    4.50    76593      4.75   
7954                    Bloons TD 5       FAMILY    4.60   190086     94.00   
7633        Five Nights at Freddy's         GAME    4.60   100805     50.00   
6746     Card Wars - Adventure Time       FAMILY    4.30   129603     23.00   

      Installs  Type  Price Content_Rating                     Genres  \
9220  10000000  Paid   6.99   Everyone 10+  Arcade;Action & Adventure   
8825  10000000  Paid   0.99     Mature 17+                     Action   
7151   1000000  Paid   6.99     Mature 17+                     Action   
7477   1000000  Paid   5.99       Everyone                Photography   
7977   1000000  Paid   5.99       Everyone                  Lifestyle   
6594   1000000  Paid   4.99       Everyone                     Action   
6082    500000  Paid   5.99       Everyone                    Weather   
7954   1000000  Paid   2.99       Everyone                   Strategy   
7633   1000000  Paid   2.99           Teen                     Action   
6746   1000000  Paid   2.99   Everyone 10+    Card;Action & Adventure   

      Revenue_Estimate  
9220     69,900,000.00  
8825      9,900,000.00  
7151      6,990,000.00  
7477      5,990,000.00  
7977      5,990,000.00  
6594      4,990,000.00  
6082      2,995,000.00  
7954      2,990,000.00  
7633      2,990,000.00  
6746      2,990,000.00

This generously assumes of course that all the installs would have been made at the listed price, which is unlikely, as there are always promotions and free give-aways on the App Stores.

The top spot of the highest-grossing paid app goes to … Minecraft at close to $70 million. It’s quite interesting that Minecraft (along with Bloons and Card Wars) is actually listed in the Family category rather than in the Game category. If we include these titles, we see that 7 out the top 10 highest-grossing apps are games. The Google Play Store seems to be quite flexible with its category labels.

# Plotly Bar Charts & Scatter Plots: Analysing App Categories

First, let's find the number of unique categories

In [29]:
df_apps_clean.Category.nunique()

33

In [30]:
top10_category = df_apps_clean.Category.value_counts()[:10]
top10_category

FAMILY             1606
GAME                910
TOOLS               719
PRODUCTIVITY        301
PERSONALIZATION     298
LIFESTYLE           297
FINANCE             296
MEDICAL             292
PHOTOGRAPHY         263
BUSINESS            262
Name: Category, dtype: int64

Based on the number of apps, the Family and Game categories are the most competitive. Releasing yet another app into these categories will make it hard to get noticed.

### Vertical Bar Chart - Highest Competition (Number of Apps)

In [31]:
bar = px.bar(x = top10_category.index, 
             y = top10_category.values)
bar.show()

### Horizontal Bar Chart - Most Popular Categories (Highest Downloads)

In [32]:
category_installs = df_apps_clean.groupby('Category').agg({'Installs': pd.Series.sum})
category_installs.sort_values('Installs', ascending=True, inplace=True)

In [33]:
h_bar = px.bar(x = category_installs.Installs,
               y = category_installs.index,
               orientation='h',
               title='Category Popularity')
 
h_bar.update_layout(xaxis_title='Number of Downloads', yaxis_title='Category')
h_bar.show()

Games and Tools are by far the most populat categories!

### Category Concentration - Downloads vs. Competition


In [37]:
category_apps_installs = df_apps_clean.groupby('Category').agg({ "App":pd.Series.count,  'Installs': pd.Series.sum})
category_apps_installs.sort_values("Installs", ascending = False)

App     Installs
Category                              
GAME                  910  13858762717
COMMUNICATION         257  11039241530
TOOLS                 719   8099724500
PRODUCTIVITY          301   5788070180
SOCIAL                203   5487841475
PHOTOGRAPHY           263   4649143130
FAMILY               1606   4437554490
VIDEO_PLAYERS         148   3916897200
TRAVEL_AND_LOCAL      187   2894859300
NEWS_AND_MAGAZINES    204   2369110650
ENTERTAINMENT         102   2113660000
BOOKS_AND_REFERENCE   169   1665791655
PERSONALIZATION       298   1532352930
SHOPPING              180   1400331540
HEALTH_AND_FITNESS    243   1134006220
SPORTS                260   1096431465
BUSINESS              262    692018120
LIFESTYLE             297    503611120
MAPS_AND_NAVIGATION   118    503267560
FINANCE               296    455249400
WEATHER                72    361096500
EDUCATION             118    352852000
FOOD_AND_DRINK         94    211677750
DATING                134    140912410
ART_AND_DESIGN         61    114233100
HOUSE_AND_HOME         62     97082000
AUTO_AND_VEHICLES      73     53129800
LIBRARIES_AND_DEMO     64     52083000
COMICS                 54     44931100
MEDICAL               292     39162676
PARENTING              50     31116110
BEAUTY                 42     26916200
EVENTS                 45     15949410

In [49]:
fig = px.scatter(category_apps_installs, 
                 x="App",
                 y="Installs",
                 size="App", 
                 color = "Installs", 
                 hover_name =category_apps_installs.index )
fig.update_layout(title = "Number of Apps (Lower = more Concentrated)", yaxis=dict(type='log') )
fig.show()

What we see is that the categories like Family, Tools, and Game have many different apps sharing a high number of downloads. But for the categories like video players and entertainment, all the downloads are concentrated in very few apps.



# Extracting Nested Data from a Column

We will answer those questions below: 
- How many different types of genres are there? Can an app belong to more than one genre? 
- Check what happens when you use .value_counts() on a column with nested values? and work around this problem by using the .split() function and the DataFrame's [.stack() method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.stack.html). 


In [51]:
#number of genres
df_apps_clean.Genres.nunique()

114

In [52]:
df_apps_clean.Genres.value_counts().sort_values(ascending = True)[:5]

Lifestyle;Pretend Play      1
Strategy;Education          1
Adventure;Education         1
Role Playing;Brain Games    1
Tools;Education             1
Name: Genres, dtype: int64

some of the apps they have multiple genres, we must split these down, then stack them

In [58]:
stack = df_apps_clean.Genres.str.split(';', expand=True).stack()
print(f'We now have a single column with shape: {stack.shape}')
num_genres = stack.value_counts()
print(f'Number of genres: {len(num_genres)}')
num_genres.head()

We now have a single column with shape: (8564,)
Number of genres: 53


Tools            719
Education        587
Entertainment    498
Action           304
Productivity     301
dtype: int64

# Colour Scales in Plotly Charts - Competition in Genres

In [55]:
bar = px.bar(x = num_genres.index[:15], # index = category name
             y = num_genres.values[:15], # count
             title='Top Genres',
             hover_name=num_genres.index[:15],
             color=num_genres.values[:15],
             color_continuous_scale='Agsunset')
 
bar.update_layout(xaxis_title='Genre',
yaxis_title='Number of Apps',
coloraxis_showscale=False)
 
bar.show()

# Grouped Bar Charts: Free vs. Paid Apps per Category

In [59]:
df_free_vs_paid = df_apps_clean.groupby(["Category", "Type"], as_index=False).agg({'App': pd.Series.count})
df_free_vs_paid.head()

Category  Type  App
0     ART_AND_DESIGN  Free   58
1     ART_AND_DESIGN  Paid    3
2  AUTO_AND_VEHICLES  Free   72
3  AUTO_AND_VEHICLES  Paid    1
4             BEAUTY  Free   42

In [61]:
g_bar = px.bar(df_free_vs_paid,
               x='Category',
               y='App',
               title='Free vs Paid Apps by Category',
               color='Type',
               barmode='group')
 
g_bar.update_layout(xaxis_title='Category',
                    yaxis_title='Number of Apps',
                    xaxis={'categoryorder':'total descending'},
                    yaxis=dict(type='log'))
 
g_bar.show()

## Create Box Plots for the Number of Installs

In [63]:
box = px.box(df_apps_clean,
             y='Installs',
             x='Type',
             color='Type',
             notched=True,
             points='all',
             title='How Many Downloads are Paid Apps Giving Up?')
 
box.update_layout(yaxis=dict(type='log'))
 
box.show()

# Plotly Box Plots: Revenue by App Category



In [64]:
df_paid_apps = df_apps_clean[df_apps_clean['Type'] == 'Paid']
box = px.box(df_paid_apps, 
             x='Category', 
             y='Revenue_Estimate',
             title='How Much Can Paid Apps Earn?')
 
box.update_layout(xaxis_title='Category',
                  yaxis_title='Paid App Ballpark Revenue',
                  xaxis={'categoryorder':'min ascending'},
                  yaxis=dict(type='log'))
 
 
box.show()